### DISCUSSION TITLE
Looking for feedback for my simple reusable component

### DISCUSSION THREAD

**Discussion by tparvi** [react-hook-form/react-hook-form#5738] [Q&A] [answered]
Relevant small codesandbox: https://codesandbox.io/s/rhf-generic-text-input-use-controller-4jxom

The sandbox has generic and non generic version of the component so you can see how they differ.

I am trying to build generic component on top of 3rd party component and **any feedback is highly appreciated**. Hopefully someone else finds this information useful as well. The code has generic and non-generic implementation and generic one has some issues.

# Issues

I should most likely use `useController` but I need to pass in T which means that it can return any type that my form has. If I don't use generics then I get `any`.

The 3rd party `TextInput` wants value to be `string|number|undefined` so TS will give following error:

```
Types of property 'value' are incompatible.
 Type 'string | number | PathValue<T, Path<T> & undefined> | PathValue<T, Path<T> & string>' is not assignable to type 'TextInputValue'.
  Type 'PathValue<T, Path<T> & undefined>' is not assignable to type 'TextInputValue'.
```

There is a hack `HACK_valueAsStringOrNumber` to get around this. **The non-generic version does not need this so any pointers on how to get rid of this hack are highly appreciated**.

Also **the generic types seems wrong** as the useWatch returns `const w: PathValue<T, undefined & Path<T>> | PathValue<T, string & Path<T>` even though it should return just `string`?

I still don't know how to display field specific error(s). How do I access them inside my generic component? WIth formik there is hook `useField<T>` but couldn't find similar concept.

# Performance

I am trying to avoid re-renders (we have formik and nested FieldArrays are slow even with v3.0 beta) so any pointers are worth of gold.

# Rendering readonly view

I am using `useWatch` to get the field value. Is this ok?

```
  const w = useWatch({ name: name, control: control });
  if (mode === "view") {
    return <div>Value is: {w}</div>;
  }
```

# Extensibility

I might have custom things like:

- custom onChange/onBlur

I guess I could add those quite easily

# TS support

Current implementation has TS support for `name` so if you write something else than `firstName` or `lastName` you get an error.

**Comment by tparvi** [comment]
I came up with simpler solution using `useController`. Unfortunately it has some of the TS issues. 
Codesandbox: https://codesandbox.io/s/rhf-generic-text-input-use-controller-4jxom

**Note the usage `any`** of `UseControllerProps<any>` **because I don't know what `any` should be**. If I use just `T` then the value type becomes `PathValue<T, Path<T>>` which seems incorrect.

```
//TextInputProps comes from 3rd party component
type Name3Props = TextInputProps & {
  mode: "edit" | "view";
};

function Name3(props: Name3Props & UseControllerProps<any>) {
  const controller = useController(props);
  const field = controller.field;

  // field.value is any but it should be string
  const temp = field.value;

  if (props.mode === "view") {
    return <span>{field.value}</span>;
  }

  return <TextInput {...props} {...field} />;
}
```

**Another issue is that when I try to use my component TS does not give error if the name property contains something that is not in ConnectionDat**a. Example:

```
// Here I specify the type for form
const { handleSubmit, control } = useForm<ConnectionData>({

// Later I use the field giving name foo and I get no TS error
<Name3 control={control} labelText="Name3" mode="edit" name="foo" />

// If I use field like this then TS gives me error because foo is
// not a property of ConnectionData 
<Controller
  name="foo"
  control={control}
  render={({ field }) => (
    <TextInput labelText="Normal usage" {...field} />
  )}
/>  
```

**Reply by bluebill1049** [reply]
for input name: `string` you probably want to use this type instead: https://react-hook-form.com/ts#FieldPath

**Reply by tparvi** [reply]
I probably misunderstood where/how to use that. Here is codesandbox: https://codesandbox.io/s/rhf-generic-text-input-fieldpath-tcbwd?file=/src/App.tsx

I changed Name3Props to generic and using `FieldPath<T>`

```
type Name3Props<T> = TextInputProps & {
  name: FieldPath<T>;
  mode: "edit" | "view";
};
```

And actual component signature to this:

```
function Name3<T>(props: Name3Props<T> & UseControllerProps<any>) {
  const controller = useController(props);
```

But it has the same issue:

Inside my component TS says `field.value` is `PathValue<any, string & Path<T>>...` even though runtime it is string:

![devtools](https://user-images.githubusercontent.com/871053/123578727-154d5300-d7df-11eb-979b-d1abe1170c99.jpg)

Am I using `useController` incorrectly? I cannot use `useController<string>` and I don't know what to put there.

**Comment by tparvi** [comment, marked as answer]
The issue about having strongly typed name is solved here https://github.com/react-hook-form/react-hook-form/discussions/5541
